src/org/zalando/stups/mint/worker/external/s3.clj

```diff
@@ -35,50 +32,51 @@
                    (.setContentLength (count bytes))
                    (.setContentType "application/json"))
         request  (doto (PutObjectRequest. bucket-name path stream metadata)
-                  (.withCannedAcl CannedAccessControlList/BucketOwnerFullControl))]
+                   (.withCannedAcl CannedAccessControlList/BucketOwnerFullControl))]
     ; to get rid of the S3V4AuthErrorRetryStrategy warnings
     (when region
       (.setRegion s3client region))
     (.putObject s3client request)))
 
-(defn writable?
-  [bucket-name app-id]
-  {:pre [(not (str/blank? bucket-name))
-         (not (str/blank? app-id))]}
-  (try
-    (put-string bucket-name
-                (str app-id "/test-mint-write")
-                {:status "SUCCESS"})
-    (log/debug "S3 bucket %s with prefix %s is writable" bucket-name app-id)
-    true
-    (catch AmazonServiceException e
-      (log/debug "S3 bucket %s with prefix %s is NOT WRITABLE. Reason %s." bucket-name app-id (str e))
-      false)))
+(defrecord S3 []
+  BucketStorage
+  (writable? [_ bucket-name app-id]
+    {:pre [(not (str/blank? bucket-name))
+           (not (str/blank? app-id))]}
+    (try
+      (put-string bucket-name
+                  (str app-id "/test-mint-write")
+                  {:status "SUCCESS"})
+      (log/debug "S3 bucket %s with prefix %s is writable" bucket-name app-id)
+      true
+      (catch AmazonServiceException e
+        (log/debug "S3 bucket %s with prefix %s is NOT WRITABLE. Reason %s." bucket-name app-id (str e))
+        false)))
 
-(defn save-user [bucket-name app-id username password]
-  {:pre [(not (str/blank? bucket-name))
-         (not (str/blank? app-id))]}
-  (try
-    (put-string bucket-name
-                (str app-id "/user.json")
-                {:application_username username
-                 :application_password password})
-    (catch AmazonServiceException e
-      (S3Exception (.getMessage e)
-                   {:status (.getStatusCode e)
-                    :message (.getMessage e)
-                    :original e}))))
+  (save-user [_ bucket-name app-id username password]
+    {:pre [(not (str/blank? bucket-name))
+           (not (str/blank? app-id))]}
+    (try
+      (put-string bucket-name
+                  (str app-id "/user.json")
+                  {:application_username username
+                   :application_password password})
+      (catch AmazonServiceException e
+        (StorageException (.getMessage e)
```

Comment:
Shouldn't here be `(throw (StorageException ...))`? `StorageException` just creates an exception, it does not throw it.